sway/desktop/layer_shell.c

```diff
@@ -293,372 +138,282 @@ void arrange_layers(struct sway_output *output) {
+	wlr_scene_node_destroy(&layer->scene->tree->node);
+}
+
+static void handle_node_destroy(struct wl_listener *listener, void *data) {
+	struct sway_layer_surface *layer =
+		wl_container_of(listener, layer, node_destroy);
+
+	// destroy the scene descriptor straight away if it exists, otherwise
+	// we will try to reflow still considering the destroyed node.
+	scene_descriptor_destroy(&layer->tree->node, SWAY_SCENE_DESC_LAYER_SHELL);
+
 	// Determine if this layer is being used by an exclusive client. If it is,
 	// try and find another layer owned by this client to pass focus to.
 	struct sway_seat *seat = input_manager_get_default_seat();
 	struct wl_client *client =
-		wl_resource_get_client(sway_layer->layer_surface->resource);
-	bool set_focus = seat->exclusive_client == client;
-
-	if (set_focus) {
-		struct sway_layer_surface *layer =
-			find_mapped_layer_by_client(client, sway_layer->layer_surface->output);
-		if (layer) {
-			seat_set_focus_layer(seat, layer->layer_surface);
+		wl_resource_get_client(layer->layer_surface->resource);
+	if (!server.session_lock.lock) {
+		struct sway_layer_surface *consider_layer =
+			find_mapped_layer_by_client(client, layer->output);
+		if (consider_layer) {
+			seat_set_focus_layer(seat, consider_layer->layer_surface);
 		}
 	}
 
-	wlr_layer_surface_v1_destroy(sway_layer->layer_surface);
-}
-
-static void handle_surface_commit(struct wl_listener *listener, void *data) {
-	struct sway_layer_surface *layer =
-		wl_container_of(listener, layer, surface_commit);
-	struct wlr_layer_surface_v1 *layer_surface = layer->layer_surface;
-	struct wlr_output *wlr_output = layer_surface->output;
-	sway_assert(wlr_output, "wlr_layer_surface_v1 has null output");
-	struct sway_output *output = wlr_output->data;
-	struct wlr_box old_extent = layer->extent;
-
-	// Rerender the static blur on change
-	if (layer->layer == ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND
-			|| layer->layer == ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM) {
-		struct fx_effect_framebuffers *effect_fbos =
-			fx_effect_framebuffers_try_get(output->wlr_output);
-		effect_fbos->blur_buffer_dirty = true;
-	}
-
-	bool layer_changed = false;
-	if (layer_surface->current.committed != 0
-			|| layer->mapped != layer_surface->surface->mapped) {
-		layer->mapped = layer_surface->surface->mapped;
-		layer_changed = layer->layer != layer_surface->current.layer;
-		if (layer_changed) {
-			wl_list_remove(&layer->link);
-			wl_list_insert(&output->layers[layer_surface->current.layer],
-				&layer->link);
-			layer->layer = layer_surface->current.layer;
-			layer_parse_criteria(layer);
+	struct wlr_layer_surface_v1 *wlr_layer_surface = layer->layer_surface;
+	if (wlr_layer_surface->current.layer ==
+			ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND ||
+		wlr_layer_surface->current.layer == ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM) {
+		if (layer->output) {
+			wlr_scene_optimized_blur_mark_dirty(root->root_scene,
+					layer->output->layers.blur_layer,
+					layer->output->wlr_output);
```

Comment:
this looks similar to the block in handle_surface_commit. I think we can move this into a function